### ISSUE TITLE
Terminology clarity - "Business microservice"

### ISSUE THREAD

**Issue by brandonwittwer** [dotnet/docs#5278]
The term "Business microservice" opens a door for the creation of mini-monoliths.  The multiple services and databases that could then live in this appropriately-cohesive set of services are no longer eligible to be independently deployed nor scaled.  Therefore it is no longer possible to achieve the important part: "The important point is that a business microservice or Bounded Context must be autonomous by allowing code and state to be independently versioned, deployed, and scaled."  The mini-monolith is now bound by some scaleable and other unscaleable components, and each must all stay in version lock step with one-another.  The only thing gained is technology choice for each service, but I feel like it comes at too great a cost and opens too large a door for mis-use.

Edit: I should mention: I'm a novice in microservice design.  I may simply be too naive and over-applying principles.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: a5660004-abca-1957-49f5-fc40f25f53dd
* Version Independent ID: 6cfd5bc4-b98e-4a8b-19c1-81afcf3f30d2
* Content: [Logical architecture versus physical architecture](https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/architect-microservice-container-applications/logical-versus-physical-architecture)
* Content Source: [docs/standard/microservices-architecture/architect-microservice-container-applications/logical-versus-physical-architecture.md](https://github.com/dotnet/docs/blob/master/docs/standard/microservices-architecture/architect-microservice-container-applications/logical-versus-physical-architecture.md)
* Service: **dotnet**
* GitHub Login: @CESARDELATORRE
* Microsoft Alias: **wiwagn**

**Comment by CESARDELATORRE** [comment]
@brandonwittwer - This is, indeed, a great question. However, in this case you are over-applying some principles over others.
A "Business microservice" is almost the same concept than the BOUNDED-CONTEXT in Domain-Driven Design. It is s boundary around a single Domain Model with all the artifacts that you need around that Domain Model.
Usually, for most of the microservices you might have just a Domain Model (with a related database) and a HTTP service, like an ASP.NET Core Web API. In that case the logical and physical implementations are coinciding.
However, that doesn't mean that it has to be like that in all the cases. For instance, in eShopOnContaines, most of the microservices are a single Web API with its related entity model and database, but the ORDERING "business microservice" or BOUNDED-CONTEXT has a different physical implementation that goes further than a single Web API because around that same Domain Model (Orderning) it needs the following:
- Web API (classic microservice)
- Host process to hold IHostedServices for background tasks (the grace period for orders)
- A SignalR Hub to handle real-time notifications when the orders change the state (submitted, paid, etc.)

The BOUNDED-CONTEXT in that case is very limited, it is just about the ORDERING subsystem/context, but in certain cases like this you need several physical services so you can focus on a different technical implementation and able to scale-out differently. For instance, the SignalR Hub could be scaled-out differently than the Web API or the host for background processes. 

In summary, the physical architecture implementation (deployment) doesn't always have to match the logical architecture. Sometimes it is better if it doesn't.
The key point here is to still limit the BOUNDED-CONTEXT to a single Domain Model which focuses on a single cohesive context, like ORDERING, CATALOG, etc.

Below you can see how in eShopOnContainers most of the microservices are composed by a single Web API except the Ordering "business microservice" or BOUNDED-CONTEXT:
![image](https://user-images.githubusercontent.com/1712635/41001497-e7224192-68c5-11e8-811a-fee99d5fcc7e.png)

Hope that helps with the question which is a really good question. 👍

**Comment by brandonwittwer** [comment]
Thank you for such a well thought out answer.  I understand the perspective, though I still question the term.  Is the intent to suggest that "Business Microservice" is the only type of "Microservice" and is simply synonymous with: "a well constrained Bounded Context".  If so: why introduce the word "Business" to the term and instead just focus on the idea that a Microservice doesn't necessarily mean only one kind of service lives within, and instead multiple types of things may possibly be running (like a web API alongside an elastic search and signalR).  

In my mind:...

I postulate that it's fair to say: even if two services share a common bounded context they don't both necessarily have to come from the same db and might very well be two identical or slightly tweaked copies of the same database and running entirely separate for one reason or another (say one has 100x load from the other, and therefore scaled independently).  

Such a layout would be perfectly acceptable and results in two smaller Microservices within a bounded context.  This could even be desirable because now each service can be versioned independently from the other and the API surface area of each is reduced from what would've otherwise  been a larger single microservice with multiple processes running.

Does that make sense?  Bounding context remains bounding context and we simply make decisions on how large of a surface area we want to deploy and version as an atomic unit on a case by case basis?

Thanks again for engaging with me on this.  I appreciate the perspective that these docs have given me and also for your experience and studied work in putting them together.

**Comment by CESARDELATORRE** [comment]
I wouldn't necessarily agree on that. A Bounded-Context has, by definition, a single Domain Model (usually backed by a database) shared by any artifacts you might need. That's what I mean.
In the moment you have a different domain model / database for another microservice, that means it is completely autonomous in regards development and deployment so it would be a different BOUNDED-CONTEXT or "business microservice".
The artifacts you have within a single Bounded-Context or business microservice have dependencies between them, that's why it is the same context within that boundary. Those elements are cohesive and have dependencies.
Coming back to the ORDERING Bounded-Context or "business microservice" (I use the word "business" because the key part here is ORDERING, which is a business concept), for the Orders I have a single domain model but I need several physical artifacts around that same small Domain Model: the Web API, a SignalR Hub and a host process to host background tasks implemented as IHostedServices.
Again, many times the logical and physical deployment implementation of a microservice can coincide (1 business microservice == a Web API) but there are other cases, like the one I mentioned for ORDERING where you need more physical artifacts around the same Domain Model and they are acting exactly within the same bounded-context or boundary. Bounded-contexts are cohesive internally, around a single Domain Model.
If you identify something that can be independent with a different versioning and autonomous deployment including its data, then, that is great, you should segregate that, but that would be a different Bounded-Context and different "Business Microservice".

**Comment by brandonwittwer** [comment]
My perception (and I assume the perception of others that read these documents) is that the addition of the word "Business" adds a new meaning to Microservice.  If I understand you correctly however you're using it only because the bounded context encompasses a business concept.  So "orders" is a business Microservice but "notifications" or "user authentication" would just be a "Microservice".  If this is what you're trying to get across I think it needs to be explicitly mentioned or just don't add the adjective at all as it just causes me confusion.

Finally, having heard the idea that a Microservice equates to a bounded context in all the ways that you have described it, I like to recognize the bounded context as more of a logical grouping of services that all share ubiquitous language and serves a single domain.  I think it's best to try to explain my opinion with the order context.  In your definition there is a single microservice that handles all things "order".  As such it contains a single domain model.  One task with orders is reviewing and searching old orders. And therefore perhaps there is a reason to create a read-optimized dataset that is purpose built for quickly loading a single or multiple orders to surface in an Order History UI.  This kind of functionality is definitely within the "order" bounded context but the data is retrieved not from a SQL server but perhaps a Hadoop cluster and a cosmosDb.  This service is independently deployable, scalable, and versionable and yet shares the conceptual model 100% with the services that take and validate orders.  

I would want to call that the OrderHistory service and I wouldn't want to say it's in it's own "Order history" bounded context because it's not distinct enough from the order context.

**Comment by brandonwittwer** [comment]
Also, I apologize if this is not the type of conversation you would like to be having in this venue and would be glad to reach out via a hangout if that would be more approproate.

**Comment by brandonwittwer** [comment]
I agree it's no exact science.  Contexts can grow complex enough to decompose into multiple contexts.  And perhaps my usecase would natuarlly stay within one context, or grow independently into it's own neighborhood of services and domain.  In your definition: The order history service, with it's own read only datastore and services still remains within the "order" microservice and is therefore versioning and deployed together..  In my definition it is a separate microservice which shares the same domain and therefore bounded-context, but splits physical existence and functionality to remain independently deployable and versionable.  I agree that it would be rare that the one service would be deployed while another isn't (any change to the data schema).  But that doesn't force me to make them required to always be a conjoined singular deployable unit.

**Comment by brandonwittwer** [comment]
Either way, I think one good thing that's come from this conversation is a better understanding of what was meant by "business" Microservice in that it simply contained a business concept.  And that another  service that did authentication wouldn't have gotten the moniker.  That would be good to clarify in the article, as I thought that "business microservice" meant something much more than "Microservice that encompasses a business process"

**Comment by mvelosop** [comment]
@mairaw I think this issue can be closed now, as it looks like the user got @CESARDELATORRE's clarification and there's nothing else pending.

**Comment by mairaw** [comment]
You should be able to close issues too @mvelosop, but please let me know if you can't.

**Comment by IAMtheIAM** [comment]
Very helpful discussion, thank you.
